### PR TITLE
Move repl to a seperate repo

### PR DESCRIPTION
**Pull request by dhoomakethu**
* REPL code moved here https://github.com/pymodbus-dev/repl
* `pymodbus.console` and `pymodbus.server` are still available once installed as before
* Installation of optional `repl` feature is still the same `pip install ".[repl]"`

This way we can keep `pymodbus` core library separate from all the helpers and utils.

## TODO
* Tests and code checks on `repl` repo to be used
* `console` and `server` commands use different approach for `commands` and `hints`, to be moved to one common lib.

**NOTE**

This is just an idea, we can discuss the best coure of action and take action with this PR accordingly,

### CODE REVIEW

**Review by janiversen — CHANGES_REQUESTED** [review]
I think you need to update doc and the README as well.

**Comment by janiversen** [comment]
Seems at least one test is also affected

ImportError while importing test module '/home/runner/work/pymodbus/pymodbus/test/sub_client/test_repl_client.py'.

**Comment by janiversen** [comment]
I think moving pymodbus.server/client to the new repo is better than having a third directory, but I am open to ideas.

File: pyproject.toml [hunk 1]
```diff
@@ -40,22 +40,16 @@ Homepage = "https://github.com/pymodbus-dev/pymodbus/"
...
-"pymodbus.console" = "pymodbus.repl.client.main:main"
-"pymodbus.server" = "pymodbus.repl.server.main:app"
 "pymodbus.simulator" = "pymodbus.server.simulator.main:main"
 
 [project.optional-dependencies]
 serial = [
     "pyserial>=3.5"
 ]
 repl = [
-    "aiohttp>=3.8.6;python_version<'3.12'",
```
[the remark this answers is not in the document]
Comment (dhoomakethu) [reply pyproject.toml:52]: OK, if simulator is not really dependent on repl, we can move this requirement to `pymodbus` as an optional dependency.

**Comment by alexrudd2** [comment]
Interesting idea.

Would it make more sense for REPL to be a separate package that depends on `pymodbus`?  That way it could stay on a stable version until newer ones have been tested.

**Comment by dhoomakethu** [comment]
> I think moving pymodbus.server/client to the new repo is better than having a third directory, but I am open to ideas.

@janiversen are you suggesting move the entire server and client to a new repo or are you talking about the REPL server and client ?

**Comment by dhoomakethu** [comment]
> Interesting idea.
> 
> Would it make more sense for REPL to be a separate package that depends on `pymodbus`? That way it could stay on a stable version until newer ones have been tested.

The idea would be to have it installed as stand-alone package and also as part of `pymodbus` library on demand, we can see if it makes sense to have it as a seperate repo instead of a subpackage to `pymodbus`.

**Comment by dhoomakethu** [comment]
> As far as I can judge, this PR remove REPL client/server from the pymodbus repo...so please explain how you where/how you would add it, because it really seems we are thinking in different terms (and my thinking might very well be wrong).

If you check the `pyproject.toml` in the PR, when somebody uses  `pip install pymodbus[repl]` or `pip install ".[repl]"` or `pip install ".[all]"`,  PIP will pull the source from https://github.com/pymodbus-dev/repl repo and install it in the `pymodbus` tree
```
repl = [
    "aiohttp>=3.8.6;python_version<'3.12'",
    "aiohttp>=3.9.0b0;python_version=='3.12'",
    "pymodbus-repl@git+https://github.com/pymodbus-dev/repl"
]
```

`pymodbus` is a `dev` dependency in `repl` project (seperate repo) and the folder structure follows packing subpackages  (`repl`) for a namespace package (in this case `pymodbus`) .
image


For more info on namespace packages refer. 
https://packaging.python.org/en/latest/guides/packaging-namespace-packages/

Now, the good thing about this is 
1. we are moving the `repl` source code away from the code pymodbus library and hence its one head-ache less for people like you who are actively maintaining it.
2. `repl` can now be installed as a subpackage to pymodbus when required from a different repo.
3. `repl` can also be installed as a seperate package altogether with `pymodbus` as a dependency. Right now it requires the latest and greatest version of pymodbus.

For 3. above,  There is no seperate PIP package available as of now and it needs to be installed from `github` directly. We can think of creating the distro depending on the demand.

**Comment by dhoomakethu** [comment]
> > > Would it make more sense for REPL to be a separate package that depends on `pymodbus`? That way it could stay on a stable version until newer ones have been tested.
> 
> Having it as a new repo that depends on a specific version of pymodbus does exactly that.
> 
> @dhoomakethu my idea was to move REPL client/server packages to a separate repo, that depends on pymodbus. That way updates to pymodbus does not affect the REPL packages, and REPL can have it own release schedule.
> 
> > The idea would be to have it installed as stand-alone package and also as part of `pymodbus` library on demand, we can see if it makes sense to have it as a seperate repo instead of a subpackage to `pymodbus`.
> 
> The focus is to make the REPL package resistent to changes in pymodbus !! I am not sure how you can do that if it is one repo. However I do not have a lot of experience in this, but I see the following problems:
> 
> * Github have only 1 release tree for each repo
> * REPL and pymodbus would at least share pyproject.toml, .github, readme etc...
> * REPL would also always use the newest pymodbus, so it would not be resistent to pymodbus changes
>   So in short, if I have understood sub-package correctly in terms of GitHub, I do not see the big difference between having a seperate repo or a sub-package
> 
> Remark: if you mean subproject, then that is a separate repo, just present in the pymodbus tree. But that would mean all pymodbus changes would affect REPL. If a new REPL repo have pymodbus as a subproject, it decides when to use a new pymodbus version.

Please refer the comment above https://github.com/pymodbus-dev/pymodbus/pull/2009#issuecomment-1942948549

**Review by alexrudd2 — APPROVED** [review]
I do not fully understand git submodules, but think @dhoomakethu has come up with a plan that considers everything.  In any case, we can figure it out together

**Comment by alexrudd2** [comment]
side note: With Chicago, Grenada and Bengaluru we have split the world timezones pretty well so that something is always happening in pymodbus :)

**Comment by janiversen** [comment]
@dhoomakethu thanks for the explanation, the missing piece for me was py project.toml, because I was merely thinking about github.

I will check it in a minute.

File: pyproject.toml [hunk 2]
```diff
@@ -40,8 +40,6 @@ Homepage = "https://github.com/pymodbus-dev/pymodbus/"
 Discord = "https://discord.gg/vcP8qAz2"
 
 [project.scripts]
-"pymodbus.console" = "pymodbus.repl.client.main:main"
```
Comment (janiversen) [inline pyproject.toml:43]: Thinking about your explanation, would it not be possible to leave this in, by having 2 small mains: repl_client.py and repl_server.py that loads the repl library conditionally (like we do with pyserial) and calls the repl package ?

File: pyproject.toml [hunk 3]
```diff
@@ -51,11 +49,9 @@ serial = [
 repl = [
     "aiohttp>=3.8.6;python_version<'3.12'",
     "aiohttp>=3.9.0b0;python_version=='3.12'",
-    "typer[all]>=0.7.0",
-    "prompt-toolkit>=3.0.36",
-    "pygments>=2.15.0",
-    "click>=8.0.0"
+    "pymodbus-repl@git+https://github.com/pymodbus-dev/repl"
```
Comment (janiversen) [inline pyproject.toml:52]: Nice

**Review by janiversen — APPROVED** [review]
Changes look good, feel free to merge once CI is green.

I updated CI some days ago, and solved a pylint issue yesterday, so just rebasing on newest dev might please CI.

**Comment by janiversen** [comment]
Ahh the culprit is documentation, you need to remove repl from there.

File: pyproject.toml (hunk 2 above)
Comment (dhoomakethu) [reply pyproject.toml:43]: It is part of the `repl` repo, the command line options will be available automatically if we install the `repl` option.

**Comment by dhoomakethu** [comment]
Just noticed there are TWO PR's created for the same branch. I am not sure, how it happened. I will get everything to one PR.

**Comment by dhoomakethu** [comment]
> Ahh the culprit is documentation, you need to remove repl from there.

@janiversen The CI failure is happening for `python 3.12` and I see the `documentation` steps are skipped in other versions. Is this something known ?

**Comment by janiversen** [comment]
Yes, we only build documentation for the newest python version, mainly because the documentation have no dependencies on python versions, so doing it for all versions will just make CI slower.

Thanks for noticing.

File: pyproject.toml (hunk 2 above)
Comment (janiversen) [reply pyproject.toml:43]: AAAh I did not know that, super !!!!

**Comment by dhoomakethu** [comment]
> Yes, we only build documentation for the newest python version, mainly because the documentation have no dependencies on python versions, so doing it for all versions will just make CI slower.
> 
> Thanks for noticing.

@janiversen so is it OK if the documentation is failing ? The error looks like a generic PYTHONPATH resolution issue and not having any thing to do with the changes introduced in this branch.

**Comment by janiversen** [comment]
No it's not OK, I think you need to remove a document and a reference....

```
doc/index.rst
doc/source/REPL.rst
doc/source/simulator.rst
```
index.rst contains a reference to REPL.rst
REPL.rst should be removed
simulator.rst contains a reference to REPL


The problem is the REPL.rst have a reference into pymodbus/repl, which is removed.

**Comment by dhoomakethu** [comment]
The CI is failing with this error for documentation
```
Running Sphinx v7.2.6

Configuration error:
There is a programmable error in your configuration file:

Traceback (most recent call last):
  File "/home/runner/work/pymodbus/pymodbus/venv/lib/python3.12/site-packages/sphinx/config.py", line 358, in eval_config_file
    exec(code, namespace)  # NoQA: S102
    ^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/pymodbus/pymodbus/doc/conf.py", line 14, in <module>
    from pymodbus import __version__ as pymodbus_version
ImportError: cannot import name '__version__' from 'pymodbus' (unknown location)
```

However trying to build locally works
```
der in 🌐 pop-os in pymodbus/doc on  repl-as-plugin [⇡] via 🐍 v3.10.12 (pymodbus) took 12s
❯ ./build_html
~/repo/pymodbus ~/repo/pymodbus/doc
updating: examples/__init__.py (stored 0%)
updating: examples/certificates/ (stored 0%)
updating: examples/client_async.py (deflated 69%)
updating: examples/client_async_calls.py (deflated 76%)
updating: examples/client_calls.py (deflated 74%)
updating: examples/client_custom_msg.py (deflated 70%)
updating: examples/client_payload.py (deflated 71%)
updating: examples/client_performance.py (deflated 71%)
updating: examples/client_sync.py (deflated 70%)
updating: examples/contrib/ (stored 0%)
updating: examples/helper.py (deflated 67%)
updating: examples/message_generator.py (deflated 78%)
updating: examples/message_parser.py (deflated 70%)
updating: examples/modbus_forwarder.py (deflated 59%)
updating: examples/server_async.py (deflated 73%)
updating: examples/server_callback.py (deflated 62%)
updating: examples/server_hook.py (deflated 62%)
updating: examples/server_payload.py (deflated 66%)
updating: examples/server_sync.py (deflated 72%)
updating: examples/server_updating.py (deflated 62%)
updating: examples/simple_async_client.py (deflated 67%)
updating: examples/simple_sync_client.py (deflated 68%)
updating: examples/simulator.py (deflated 65%)
updating: examples/datastore_simulator_share.py (deflated 67%)
~/repo/pymodbus/doc
Running Sphinx v7.2.6
loading pickled environment... done
building [mo]: targets for 0 po files that are out of date
writing output...
building [html]: targets for 0 source files that are out of date
updating environment: [config changed ('recommonmark_config')] 21 added, 0 changed, 0 removed
reading sources... [100%] source/simulator
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
copying downloadable files... [100%] source/_static/examples.tgz
copying static files... done
copying extra files... done
done
writing output... [100%] source/simulator
generating indices... genindex py-modindex done
writing additional pages... search done
copying images... [100%] source/library/simulator/modbuscommonblock.png
dumping search index in English (code: en)... done
dumping object inventory... done
build succeeded.

The HTML pages are in ../build/html/html.
```

Will check if its specific to python3.12

**Comment by dhoomakethu** [comment]
works on `python3.12-.2` locally

```
rider in 🌐 pop-os in pymodbus/doc on  repl-as-plugin [!] via 🐍 v3.12.2+ (pymodbus3.12)
❯ ./build_html
~/repo/pymodbus ~/repo/pymodbus/doc
updating: examples/__init__.py (stored 0%)
updating: examples/certificates/ (stored 0%)
updating: examples/client_async.py (deflated 69%)
updating: examples/client_async_calls.py (deflated 76%)
updating: examples/client_calls.py (deflated 74%)
updating: examples/client_custom_msg.py (deflated 70%)
updating: examples/client_payload.py (deflated 71%)
updating: examples/client_performance.py (deflated 71%)
updating: examples/client_sync.py (deflated 70%)
updating: examples/contrib/ (stored 0%)
updating: examples/helper.py (deflated 67%)
updating: examples/message_generator.py (deflated 78%)
updating: examples/message_parser.py (deflated 70%)
updating: examples/modbus_forwarder.py (deflated 59%)
updating: examples/server_async.py (deflated 73%)
updating: examples/server_callback.py (deflated 62%)
updating: examples/server_hook.py (deflated 62%)
updating: examples/server_payload.py (deflated 66%)
updating: examples/server_sync.py (deflated 72%)
updating: examples/server_updating.py (deflated 62%)
updating: examples/simple_async_client.py (deflated 67%)
updating: examples/simple_sync_client.py (deflated 68%)
updating: examples/simulator.py (deflated 65%)
updating: examples/datastore_simulator_share.py (deflated 67%)
~/repo/pymodbus/doc
Running Sphinx v7.2.6
loading pickled environment... done
building [mo]: targets for 0 po files that are out of date
writing output...
building [html]: targets for 21 source files that are out of date
updating environment: [config changed ('recommonmark_config')] 21 added, 0 changed, 0 removed
reading sources... [100%] source/simulator
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
copying downloadable files... [100%] source/_static/examples.tgz
copying static files... done
copying extra files... done
done
writing output... [100%] source/simulator
generating indices... genindex py-modindex done
writing additional pages... search done
copying images... [100%] source/library/simulator/modbuscommonblock.png
dumping search index in English (code: en)... done
dumping object inventory... done
build succeeded.

The HTML pages are in ../build/html/html.
```

**Comment by janiversen** [comment]
but it works normally as you can see on dev.

I will try to checkout your PR and see what happens.

**Comment by janiversen** [comment]
This is real strange, dev works perfectly on CI but this PR do not....something does something unexpected.

**Comment by dhoomakethu** [comment]
I will try to poke around a bit. Will try to reproduce this locally, my wild guess is this is something to do with cached env's.

**Comment by janiversen** [comment]
good catch, I did not think of that.

I just removed the aiohttp dependencies of REPL, those should be in the repl repo.

**Comment by janiversen** [comment]
If this one get green, then lets merge.

**Comment by janiversen** [comment]
This PR was one of the heavier ones, but thanks for helping making it happen.

@dhoomakethu having REPL in a separate repo, means it is primarily your child....but poke me if you need help etc ...I do not want to abandon REPL, which is a very good tool.

If you want the REPL documentation to be part of the pymodbus documentation (but with its own versioning), just know I am all in favor !

**Comment by dhoomakethu** [comment]
We will have the documentation moved to repl repo and link it in pymodbus. I will update the PR tomorrow

**Pull request merged**